src/jitlayers.cpp

```diff
@@ -2315,54 +2007,225 @@ void JuliaOJIT::enablePerfJITEventListener()
+            continue;
+        JL_GC_PROMISE_ROOTED(CI);
+        Syms.at(T)->setName(linkCallTarget(CI, API));
+    }
+
+    // Rename globals and add mappings
+    // TODO: don't leak when we have a way to GC code
+#ifdef __clang_analyzer__
+    [[clang::suppress]]
 #endif
+    void **Ptrs = new void *[Info->global_targets.size()];
+    size_t i = 0;
+    orc::SymbolMap GlobalSyms;
+    for (auto &[Addr, Orig] : Info->global_targets) {
+        auto Sym = ES.intern(Names(*Orig, "#"));
+        auto It = Syms.find(Orig);
+        if (It == Syms.end())
+            continue;
+        It->second->setName(Sym);
+        Ptrs[i] = Addr;
+        GlobalSyms[Sym] = {ExecutorAddr::fromPtr(Ptrs + i), JITSymbolFlags::Exported};
+        ++i;
+        ++LinkedGlobals;
+    }
+    cantFail(JD.define(orc::absoluteSymbols(std::move(GlobalSyms))));
 
-const DataLayout& JuliaOJIT::getDataLayout() const
+    DebuginfoPlugin->notifyMaterializingWithInfo(MR, G, ObjBuf, std::move(Info));
+}
+
+// Must hold LinkerMutex.
+orc::SymbolStringPtr JuliaOJIT::linkCallTarget(jl_code_instance_t *CI, jl_invoke_api_t API)
 {
-    return DL;
+    auto It = CISymbols.find(CI);
+    if (It != CISymbols.end())
+        return It->second.specptr;
+
+    CISymbolPtr *Sym = linkCISymbol(CI);
+
+    orc::SymbolStringPtr Result;
+    CISymbolPtr Trampoline;
+    std::unique_ptr<jl_codegen_output_t> Out;
+
+    if (!Sym) {
+        // The target CI was not compiled when we observed it, so generate a
+        // tojlinvoke trampoline that will cause it to be compiled.
+        // TODO: replace this with a GOT/PLT mechanism that avoids the jl_invoke
+        // after it has been compiled.
+        Out = std::make_unique<jl_codegen_output_t>("tojlinvoke", getDataLayout(),
+                                                    getTargetTriple());
+        Function *F = emit_tojlinvoke(CI, StringRef(), *Out);
+        Sym = &Trampoline;
+        Trampoline.invoke_api = API;
+
+        if (API == JL_INVOKE_SPECSIG) {
+            // codegen may contain safepoints (such as jl_subtype calls)
+            int8_t GCState = jl_gc_unsafe_enter(jl_current_task->ptls);
+            Function *G = emit_specsig_to_fptr1(*Out, CI, F);
+            G->setLinkage(GlobalValue::ExternalLinkage);
+            jl_gc_unsafe_leave(jl_current_task->ptls, GCState);
+            Trampoline.specptr = mangle(G->getName());
+        } else if (API == JL_INVOKE_ARGS) {
+            F->setLinkage(GlobalValue::ExternalLinkage);
+            Trampoline.invoke_api = JL_INVOKE_ARGS;
+            Trampoline.specptr = mangle(F->getName());
+        } else {
+            abort();
+        }
+    }
+
+    if (Sym && Sym->invoke_api == API) {
+        // If the CI has a compiled specptr, and this call site uses the same
+        // calling convention, link to it directly.
+        Result = Sym->specptr;
+    }
+    else {
+        // TODO: Check invariants to make sure this can't happen.
+        abort();
```

Comment:
Happens in the (unlikely) case that it predicted we would want to compile with specsig, then later the runtime changes its mind (e.g. because it later ends up inferring `@nospecialize` from intersecting methods, or `--compile=min` is in use, etc.)